### PR TITLE
Initial feature commit.

### PR DESCRIPTION
**Pull request by tparikka**
Commit for #16. 

Adds the option to provide a birthdate and have age at each goal displayed. Would love any feedback you can provide or revisions you'd like me to make before accepting this PR. Thanks for your consideration!

### CODE REVIEW

File: src/app/forecasting/models/calculate-input.model.ts [hunk 1]
```diff
@@ -8,12 +8,13 @@ export class CalculateInput {
   expectedAnnualGrowthRate = 0;
   monthlyContribution = 0;
   leanFiPercentage = 0;
+  public birthdate: Date = null;
```
Comment (JackMorrissey) [inline src/app/forecasting/models/calculate-input.model.ts]: Please remove the `public` to keep it consistent with the existing codebase. By default these variables are exposed to the html files; you explicitly mark it as private if you want to restrict it.

File: src/app/forecasting/input/ynab/ynab.component.ts [hunk 2]
```diff
@@ -24,6 +24,7 @@ export class YnabComponent implements OnInit {
   displayContributionInfo = true;
   public safeWithdrawalRatePercentage = 4.00;
   public expectedAnnualGrowthRate = 7.00;
+  public birthdate: Date;
```
[the remark this answers is not in the document]
Comment (tparikka) [reply src/app/forecasting/input/ynab/ynab.component.ts]: Far be it from me to anger the squiggly gods. Change coming in next commit. :)

File: src/app/forecasting/models/calculate-input.model.ts (hunk 1 above)
Comment (tparikka) [reply src/app/forecasting/models/calculate-input.model.ts]: Removed access modifier, thanks for the correction!

File: src/app/forecasting/input/ynab/ynab.component.html [hunk 3]
```diff
@@ -20,8 +19,8 @@
               <div class="input-group">
                 <select class="form-control" id="selectedBudget" formControlName="selectedBudget">
                   <option *ngFor="let budget of budgets"
-                    [value]="budget.id">
-                      {{budget.name}}
+                  [value]="budget.id">
```
[the remark this answers is not in the document]
Comment (tparikka) [reply src/app/forecasting/input/ynab/ynab.component.html]: Did you want the line breaks restored as well, or just the spacing on those option controls?

File: src/app/forecasting/input/ynab/ynab.component.html [hunk 4]
```diff
@@ -124,6 +122,17 @@
             </div>
           </div>
         </div>
+        <div class="form-group">
+          <div class="row">
+            <label for="birthdate" class="col-sm-12 col-md-5 col-form-label">Birth date</label>
+            <div class="col">
+              <div class="input-group">
+                <input type="date" class="form-control" id="birthdate" formControlName="birthdate" step="1"
```
[the remark this answers is not in the document]
Comment (tparikka) [reply src/app/forecasting/input/ynab/ynab.component.html]: Date picker coming in the next commit.

**Comment by JackMorrissey** [comment]
Sorry for the delay. This looks good to me. Only thing I'd like to change is making the date input able to be used without necessarily needing the calendar input. Not a big deal though. If you could rebase these commits together with the suggested commit message above we can get this in. Thanks again!

**Comment by tparikka** [comment]
That's okay, thanks for giving it a look! I have a question about making the date input independent from the calendar control though. Right now all the inputs are collected in a FormBuilder, there's a local const formChanges that receives change events emitted after a 500ms debounce timer, and when it receives an event it updates all the inputs. Except, if you don't type the date REAL fast the debounce time passes, the event triggers, and the updateInput function tries to parse a string that isn't actually a complete date. Is there a non-hacky way to check the value of that calendar control is a valid parseable date without individually checking each field in the birth date object returned by the calendar control?

**Comment by JackMorrissey** [comment]
Here's what I came up with:

```html
<input class="form-control" id="ngBirthDate" formControlName="ngBirthDate"
                  ngbDatepicker #d="ngbDatepicker" [minDate]="{year: 1900, month: 1, date: 1}" placeholder="yyyy-mm-dd">
```

```ts
    const formBirthDate = this.budgetForm.value.ngBirthDate;
    if (formBirthDate && formBirthDate.year && formBirthDate.month && formBirthDate.day) {
      const parsedBirthDate = new Date(`${this.budgetForm.value.ngBirthDate.year}-${this.budgetForm.value.ngBirthDate.month}-
          ${this.budgetForm.value.ngBirthDate.day}`);
      result.birthDate = parsedBirthDate;
      this.birthDate = parsedBirthDate;
    }
```

The trick to dodge the debounce issue is to just not set the value if it's invalid. After console.logging the output of formBirthDate, I saw that it's a string when invalid and has those properties if it parses, so I just used that as my check. I also went back to that documentation for the DatePicker and matched the example (removing the  `type="date"` on that field)

Next nice to have on top of that would be to put some validation on the input for the datepicker, but not required for the initial version.

**Pull request closed without merging**